server-spi-private/src/main/java/org/keycloak/authorization/policy/evaluation/DecisionPermissionCollector.java

```diff
@@ -42,116 +44,117 @@
  */
 public class DecisionPermissionCollector extends AbstractDecisionCollector {
 
-    private final AuthorizationProvider authorizationProvider;
-    private final ResourceServer resourceServer;
-    private final AuthorizationRequest request;
-    private final Set<Permission> permissions = new LinkedHashSet<>();
+    protected final AuthorizationProvider authorizationProvider;
+    protected final ResourceServer resourceServer;
+    protected final AuthorizationRequest request;
+    protected final Set<Permission> permissions = new LinkedHashSet<>();
 
     public DecisionPermissionCollector(AuthorizationProvider authorizationProvider, ResourceServer resourceServer, AuthorizationRequest request) {
         this.authorizationProvider = authorizationProvider;
         this.resourceServer = resourceServer;
         this.request = request;
     }
 
+    /**
+     * Processes the final permission decisions for all policies associated with a single resource permission
+     * @param results a map of {@link Policy} to {@link Result} containing all needed contextual and hierarchical information for making a decision
+     */
     @Override
-    public void onComplete(Result result) {
-        ResourcePermission permission = result.getPermission();
-        Resource resource = permission.getResource();
+    public void onComplete(ResourcePermission permission, Map<Policy, Result> results) {
         Collection<Scope> requestedScopes = permission.getScopes();
-
-        if (Effect.PERMIT.equals(result.getEffect())) {
-            if (permission.getScopes().isEmpty() && !resource.getScopes().isEmpty()) {
-                return;
+        Scope fakeScope = null;
+
+        // we need to allow requests to scopes that don't actually exist on the resource so we need to do some processing...
+        Collection<Scope> resourceScopes = new LinkedHashSet<>();
+        if(permission.getResource() != null) {
+           resourceScopes.addAll(permission.getResource().getScopes());
+            // in some cases we end up with no temp scopes to work with. We still need to be able to track overrides on these policies
+            // create a fake scope relevant to this resource that we'll remove at the end
+            fakeScope = new FakeScope(permission.getResource().getId());
+        }
+        // if we don't have a resource we can still try to grant the scopes requested outright
+        resourceScopes.addAll(requestedScopes);
+
+        // without having some modifications to how policies are queried we need filter out a bunch of stuff that shouldn't be here.
+        // for each result entry:
+        Stream<Result> resourcePermissions = results.values().stream()
```

Comment:
The contributions from #25920 will remove all of this stream filtering and DRASTICALLY improve the performance at scale of the authorization system as it removes duplicate searches and unrelated searches from the results. 

As of present, there are areas of the evaluation process that query every single policy currently in the database and evaluate them which is not performant.

Reply:
This still applies though it's labelled outdated, I just updated the filters